### PR TITLE
Fix replaceAccentedChars function

### PR DESCRIPTION
**Pull request by davidglezz**
| Questions         | Answers
| ----------------- | -------------------------------------------------------
| Branch?           | develop
| Description?      | Fix replaceAccentedChars by adding some missing accented chars.
| Type?             | bug fix
| Category?         | CO
| BC breaks?        | no
| Deprecations?     | no
| Fixed ticket?     | no
| How to test?      | Check `testReplaceAccentedChars()`
| Possible impacts? | Nah. Some chars that were not unaccented before, now are unaccented.

My test function:
```php
    public function testReplaceAccentedChars(): void
    {
        // https://en.wikipedia.org/wiki/Tilde
        // https://en.wikipedia.org/wiki/List_of_Unicode_characters
        // https://graphemica.com/
        $testStrings = [
            'aáàäâAÁÀÄÂ' => 'aaaaaAAAAA',
            'eéèëêEÉÈËÊ' => 'eeeeeEEEEE',
            'iíìïîIÍÌÏÎĨ' => 'iiiiiIIIIII',
            'oóòöôOÓÒÖÔ' => 'oooooOOOOO',
            'uúùüûUÚÙÜÛ' => 'uuuuuUUUUU',
        ];

        foreach ($testStrings as $testString => $expected) {
            $actual = Tools::replaceAccentedChars($testString);
            $this->assertEquals($expected, $actual);
        }
    }

```

Question: do we want to remove the accent or do we want to substitute with something phonetically similar?
here I just remove the accent.

```diff
--- Before this PR
+++ After this PR
@@ @@
-'aaaaaAaaAa'
+'aaaaaAAAAA'
-'eeeeeEeeEe'
+'eeeeeEEEEE'
-'iiiiiIiiÏÎi'
+'iiiiiIIIIII'
-'oooooOooÖo'
+'oooooOOOOO'
-'uuuuuUuuUU'
+'uuuuuUUUUU'
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/27176)

### CODE REVIEW

**Review by matks — COMMENTED** [review]
@davidglezz as you have some clear usecases for this function, what about we add them inside a unit test 😉  ?

See https://github.com/PrestaShop/PrestaShop/blob/develop/tests/Unit/Classes/ToolsTest.php

**Review by matthieu-rolland — COMMENTED** [review]
lgtm, unit test welcome 👍

**Review by matks — DISMISSED** [review]
Still waiting for tests 😉

**Comment by davidglezz** [comment]
Done!

**Comment by davidglezz** [comment]
I use [transliterator_transliterate()](https://www.php.net/manual/en/transliterator.transliterate.php) This functions requires `Intl` extension.

The test can be removed. It is just a sample of how it convert the chars.

```php
$str = 'キャンパス  - Αλφαβητικός Κατάλογος - биологическом';
echo transliterator_transliterate('Any-Latin; Latin-ASCII', $str);
//> kyanpasu  - Alphabetikos Katalogos - biologiceskom
```

File: classes/Tools.php [hunk 1]
```diff
@@ -1499,100 +1499,13 @@ public static function str2url($str)
...
-        // å to aa
-        // ä to ae
+        static $transliterator;
 
-        $replacements = [
-            'a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'r', 's', 'ss', 't', 'u', 'v', 'w', 'y', 'z', 'ae', 'ch', 'kh', 'oe', 'sh', 'shh', 'ya', 'ye', 'yi', 'yo', 'yu', 'zh',
-            'A', 'B', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'J', 'K', 'L', 'M', 'N', 'O', 'P', 'R', 'S', 'T', 'U', 'V', 'W', 'Y', 'Z', 'AE', 'CH', 'KH', 'OE', 'SH', 'SHH', 'YA', 'YE', 'YI', 'YO', 'YU', 'ZH',
-        ];
+        if (!$transliterator) {
+            $transliterator = transliterator_create('Any-Latin; Latin-ASCII');
```
Comment (atomiix) [inline classes/Tools.php]: Maybe use the object oriented way?
```suggestion
            $transliterator = Transliterator::create('Any-Latin; Latin-ASCII');
```

File: classes/Tools.php [hunk 2]
```diff
@@ -1499,100 +1499,13 @@ public static function str2url($str)
...
+        }
 
-        return preg_replace($patterns, $replacements, $str);
+        return transliterator_transliterate($transliterator, $str);
```
Comment (atomiix) [inline classes/Tools.php]:
```suggestion
        return $transliterator->transliterate($str);
```

**Comment by davidglezz** [comment]
@atomiix done

**Review by atomiix — APPROVED** [review]

**Comment by davidglezz** [comment]
Please remove "Waiting for author" label, or do you need something else?

**Review by Progi1984 — APPROVED** [review]

**Review by PierreRambaud — APPROVED** [review]

**Comment by AureRita** [comment]
Hi @davidglezz ,

Thank you for you pr but can you create an issue or feature to allow to the QA team to have few steps to follow to test your pr please ? :) 

that will allow us to test it quickly

**Comment by davidglezz** [comment]
This started as a simple bugfix, but now, it's so much more.
My start use case is as simple as: call `Tools::replaceAccentedChars($testString);` does not return expected string. (See the description). That's why I didn't create an issue and did a PR directly.

Well, looking for related issues I see that this PR fixes some of them:

## This PR fix:
- My basic use case
- Allow friendly URLs to turn Greek into Latin characters #21978
   - Ελένη → Elene
- Many other scenarios!:

```php
echo Tools::replaceAccentedChars('キャンパス  - Αλφαβητικός Κατάλογος - биологическом');
//> kyanpasu  - Alphabetikos Katalogos - biologiceskom
```

## Requires a decision: Locale dependent:
- Incorrect accented chars replacement for ö and ä #26157 #26160
This PR could also fix this, but it requires a decision, because they are locale-dependent transformations and might not be right for everyone.
  - https://github.com/PrestaShop/PrestaShop/issues/26157#issuecomment-1072994521

My proposal is to add a `$locale` parameter to the `replaceAccentedChars` function.
Note: there is an `de-ASCII` transliterator for `de` locale.

## Inconsistencies between js and php implementation
Can be solved in other PR. My proposal is to deprecate the current js function `str2url` and add a new one that makes a ajax call to the backend to use `Tools::str2url()` PHP function.

## Unrelated issues:
- Bug on "reset url" button of the product pages when "Accented URLs" is enabled and there are special characters in the product name #21495
  - URL Rewrite problem #20090

**Review by AureRita — DISMISSED** [review]
Hi @davidglezz 

I tested your pr and : 

I tested your first lane (with a) with and without your pr ass you can see : 
image

that was the friendly url : 
image

Moreover, I also test the issue you linked ( that why they are linked ) : 
for Ελένη that doesn't wirte "Elene", that write nothing, as you can see  : 
image

For other characters that also change nothing : 
image

**Comment by davidglezz** [comment]
@AureRita thanks for your time.
This PR is (depending on the point of view) an improvement or bugfix of `replaceAccentedChars` function.

This function is used by:
- Search & search alias
- `str2url` & `slugify`
- `smartyClassname`

I think you've tried to test it by using the `str2url` function, but doing it with the "Reset URL" button in the product page is using the similar `str2url` javascript function.

In another PR I want to modify that javascript function so that it uses the `str2url` function of the backend (PHP) so that there is consistency and there are no more issues regarding this topic.

To avoid confusion, there are 2 `str2url` functions, a javascript one that runs in the browser and a PHP that runs on the server.

**Comment by AureRita** [comment]
hi @davidglezz, 

Thank you for your answer, Currently, I don't how to test these type of variable so can you explain to me how can i test 
Search & search alias
str2url & slugify
smartyClassname
with your pr to accept or not your pr,

**Comment by davidglezz** [comment]
I have unlinked #21978 form the description table because does not fully fixes it, but I think this is the first step to fix it.

@AureRita I think that since it is a php function, it cannot be tested directly from the user interface, I suppose it is fine if the unit tests and the automated tests pass.

**Comment by AureRita** [comment]
hi @davidglezz , 

Indeed, so we will run an automated test on it, 
can you @PrestaShop/prestashop-core-developers check for the unit test ? Thank you in advance

**Comment by AureRita** [comment]
Hi @davidglezz ,

I ran an automated test on it, and it haven't issues, now we will wait the dev team

**Comment by atomiix** [comment]
@AureRita the code has been approved with unit tests and the CI is green so I think it's okay

**Review by AureRita — APPROVED** [review]

**Comment by Progi1984** [comment]
Thanks @davidglezz & @AureRita

**Pull request merged**